### PR TITLE
Imputation time series splitter

### PR DESCRIPTION
**Pull request by tomxi**
implemented a function in the data module for generating masks for training and test split, given a time series. It is done thinking about the imputation task context.

### CODE REVIEW

File: sonyc_kalman/data.py [hunk 1]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+
+    Params:
+    -------
+        invalid_mask: list or np.array
+            an array of 1's and 0's. 0 being valid, 1 being invalid
+        test_ratio: float (0, 1)
+            the fraction of the data to be split out as test
+        mode: str
+            should be one of 'random_holes', 'r', 'chronological', or 'c'.
+            'random_holes' - punch_out randomly (Guasssian according to `hole_param`) sized holes in the series and
```
Comment (ChrisIck) [inline sonyc_kalman/data.py]: "Guassian" lol, typo?

File: sonyc_kalman/data.py [hunk 2]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+            hole_param in chronological mode is ignored.
+        hole_param: list [mean, std]
+            used for the 'random_holes' mode.
+            
+    Returns:
+    --------
+        train_mask: np.array of int's
+            an array of 1's and 0's that can be used to mask any series. 1 means selected for the training split
+        test_mask: np.array of int's
+            an array of 1's and 0's that can be used to mask any series. 1 means selected for the test split
```
Comment (ChrisIck) [inline sonyc_kalman/data.py]: Let's be consistent with our conventions. IIRC we decided on following numpy.ma conventions (1=masked, 0=readable), so we should probably switch these (unless I'm misinterpreting this). That way, we can just take our input data and apply a numpy mask to it to our data for training/evaluation. @jtcramer, let me know what you think of this.

**Review by ChrisIck — CHANGES_REQUESTED** [review]
Looks good, but we should be consistent in our mask conventions so we don't have to flip masks on every import. Up until now, I've been doing 0=valid, 1=invalid, and I think we should follow these conventions for our train/test masks so we can just apply them to our data for training and evaluation, rather than needing to flip them. Once we decide on that, though, I think this is good to push.

File: sonyc_kalman/data.py [hunk 3]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
     return X, mask
 
 
+def series_splitter(invalid_mask, test_ratio=0.25, mode='random_holes', hole_param=[20, 10]):
```
Comment (jtcramer) [inline sonyc_kalman/data.py]: Make `hole_param=None` instead of a list, and then assign `hole_param = [20, 10]` if `hole_param is None`. The way that default arguments work with mutable data-types is tricky: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: sonyc_kalman/data.py (hunk 1 above)
Comment (jtcramer) [reply sonyc_kalman/data.py]: Dat Guasssian doe

File: sonyc_kalman/data.py (hunk 2 above)
Comment (jtcramer) [reply sonyc_kalman/data.py]: I'm in favor of consistency across the board.

File: sonyc_kalman/data.py (hunk 3 above)
Comment (jtcramer) [reply sonyc_kalman/data.py]: Also we should split `hole_param` into `hole_mean` and `hole_var` rather than this list.

File: sonyc_kalman/data.py [hunk 4]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+        continuous_split_train = 1 - continuous_split_test
+        # then use the valid_mask to put these contiguous masks into the right places
+        test_mask[valid_mask]  = continuous_split_test
+        train_mask[valid_mask] = continuous_split_train
+
+    else:
+        # 'random_holes' mode
+        ## first generate an array that holes the sizes of the holes, with length n_holes
+        hole_sizes = []
+        while sum(hole_sizes) < n_test:
```
Comment (jtcramer) [inline sonyc_kalman/data.py:127]: More of an open question than an actionable item. Do we want to have a minimum hole size?

File: sonyc_kalman/data.py [hunk 5]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+        ## next, decide where the holes are gonna be, ie, how to split the training data into n_holes + 1 parts
+        n_holes = len(hole_sizes)
+        # pick n_holes positions from all n_train+1 positions: this avoids two adjacent holes
+        hole_start_positions = np.sort(np.random.choice(n_train+1, n_holes, replace=False))
+        inter_hole_intervals = [hole_start_positions[0]] + list(np.diff(hole_start_positions))
+        after_last_hole = n_train - hole_start_positions[-1]
+
+        ## then build the continuous_split arrays according to hole_sizes
+        continuous_split_test = []
+        for hole, bread in zip(hole_sizes, inter_hole_intervals):
```
Comment (jtcramer) [inline sonyc_kalman/data.py:145]: I approve of this wordplay. :)

File: sonyc_kalman/data.py [hunk 6]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+            continuous_split_test += bread * [0]
+            continuous_split_test += hole * [1]
+        continuous_split_test += after_last_hole * [0]
+
+        ## finally put things into places
+        continuous_split_test = np.array(continuous_split_test, dtype=int)
+        continuous_split_train = 1 - continuous_split_test
+        test_mask[valid_mask]  = continuous_split_test
+        train_mask[valid_mask] = continuous_split_train
+
```
Comment (jtcramer) [inline sonyc_kalman/data.py:153]: It would be good to also have a third option like what we had discussed, where we uniformly divide the year into chunks of a given time (e.g 4 weeks) and then alternate assignment (or randomly assign) between training and test sets.

File: sonyc_kalman/data.py (hunk 3 above)
Comment (jtcramer) [reply sonyc_kalman/data.py]: It'd also be good to pass in a `random_state` parameter and then set the random seed with `np.random.seed` and `random.seed` for reproducability.

File: sonyc_kalman/data.py [hunk 7]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+        # then use the valid_mask to put these contiguous masks into the right places
+        test_mask[valid_mask]  = continuous_split_test
+        train_mask[valid_mask] = continuous_split_train
+
+    else:
+        # 'random_holes' mode
+        ## first generate an array that holes the sizes of the holes, with length n_holes
+        hole_sizes = []
+        while sum(hole_sizes) < n_test:
+            rand_size = int(round(np.random.normal(hole_param[0], hole_param[1])))
```
Comment (jtcramer) [inline sonyc_kalman/data.py]: Also it might be good to also have deterministic holes so that we can evaluate performance as a function of hole size. Though as I type this I realize we can always just set the "scale" parameter to be 0.

File: sonyc_kalman/data.py [hunk 8]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+        ## then build the continuous_split arrays according to hole_sizes
+        continuous_split_test = []
+        for hole, bread in zip(hole_sizes, inter_hole_intervals):
+            continuous_split_test += bread * [0]
+            continuous_split_test += hole * [1]
+        continuous_split_test += after_last_hole * [0]
+
+        ## finally put things into places
+        continuous_split_test = np.array(continuous_split_test, dtype=int)
+        continuous_split_train = 1 - continuous_split_test
```
Comment (jtcramer) [inline sonyc_kalman/data.py]: Just clarifying for my own understanding; we only hallucinate holes for the test set right?

File: sonyc_kalman/data.py [hunk 9]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+    n_train = n_total_valid - n_test
+    
+    train_mask = valid_mask.copy()
+    test_mask = valid_mask.copy()
+    if chrono_mode:
+        # first split the data as if all the valid frames are contiguous
+        continuous_split_test = np.array([0] * n_train + [1] * n_test, dtype=int)
+        continuous_split_train = 1 - continuous_split_test
+        # then use the valid_mask to put these contiguous masks into the right places
+        test_mask[valid_mask]  = continuous_split_test
```
Comment (jtcramer) [inline sonyc_kalman/data.py]: Note that we can still use the invalid samples in the test/training set, just that they won't be used as observations. We should still include them in the train/test splits.

File: sonyc_kalman/data.py [hunk 10]
```diff
@@ -70,6 +70,93 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+        for hole, bread in zip(hole_sizes, inter_hole_intervals):
+            continuous_split_test += bread * [0]
+            continuous_split_test += hole * [1]
+        continuous_split_test += after_last_hole * [0]
+
+        ## finally put things into places
+        continuous_split_test = np.array(continuous_split_test, dtype=int)
+        continuous_split_train = 1 - continuous_split_test
+        test_mask[valid_mask]  = continuous_split_test
+        train_mask[valid_mask] = continuous_split_train
```
Comment (jtcramer) [inline sonyc_kalman/data.py]: Same as before, "Note that we can still use the invalid samples in the test/training set, just that they won't be used as observations. We should still include them in the train/test splits."

**Review by jtcramer — CHANGES_REQUESTED** [review]

File: sonyc_kalman/data.py (hunk 6 above)
Comment (jtcramer) [reply sonyc_kalman/data.py:153]: For this though, we'd have to pass in a list of timestamps

File: sonyc_kalman/data.py (hunk 3 above)
Comment (tomxi) [reply sonyc_kalman/data.py]: Not sure exactly how to do `np.random.seed`

File: sonyc_kalman/data.py (hunk 6 above)
Comment (tomxi) [reply sonyc_kalman/data.py:153]: will punt on this for now...

File: sonyc_kalman/data.py (hunk 8 above)
Comment (tomxi) [reply sonyc_kalman/data.py]: In the hole mode, the holes are the test sets... or vise versa.
The var names are a bit misleading, but since this is not a train/val/test splitter, the output masks can be used (in combination with the input mask) to split the `holes` into chronological pieces etc...

File: sonyc_kalman/data.py (hunk 9 above)
Comment (tomxi) [reply sonyc_kalman/data.py]: Ok, so for all the invalid ones, I won't mask them out. In other words, if `mask[i] == 1` then `test_mask[i] == 0, train_mask[i] == 0` right?

File: sonyc_kalman/data.py (hunk 10 above)
Comment (tomxi) [reply sonyc_kalman/data.py]: ditto.

**Comment by tomxi** [comment]
My dearest CRer, sorry for the massive delay. Shall we have another go on looking through this?

File: sonyc_kalman/data.py (hunk 8 above)
Comment (jtcramer) [reply sonyc_kalman/data.py]: Got it.

File: sonyc_kalman/data.py (hunk 3 above)
Comment (jtcramer) [reply sonyc_kalman/data.py]: You can call `np.random.seed(<int>)` to set the random seed

File: sonyc_kalman/data.py (hunk 9 above)
Comment (jtcramer) [reply sonyc_kalman/data.py]: I think what I meant was that the mask doesn't need to factor into this at all. All samples (including invalid samples) should be included in the train/test split.

A related concern: after thinking about it a bit more, _ideally_ we should only have invalid samples in the training set, and none in the test set. Though, this doesn't fundamentally change how we evaluate. We can still have invalid samples in the test set and then just measure error over valid samples.

We could also deal with boundary conditions, i.e. ensuring that we don't start or end a sequence with invalid samples. But we can punt this to later.

**Review by jtcramer — CHANGES_REQUESTED** [review]
Looks good for the most part, just consider the comments on the invalid samples masking.

File: sonyc_kalman/data.py (hunk 9 above)
Comment (tomxi) [reply sonyc_kalman/data.py]: > ideally we should only have invalid samples in the training set

This is how I'll do it. since the ratio reflect how much valid data is held-out for the test/hole partition

**Comment by tomxi** [comment]
Another round?

File: sonyc_kalman/data.py [hunk 11]
```diff
@@ -70,6 +70,105 @@ def load_openl3_time_series(hdf5_path, delta_mins=15, aggr_func=None):
...
+            continuous_split_test += bread * [1]
+            continuous_split_test += hole * [0]
+        continuous_split_test += after_last_hole * [1]
+
+        ## finally put things into places
+        continuous_split_test = np.array(continuous_split_test, dtype=int)
+        # only select regions in valid_idx for test_mask
+        test_mask[valid_idx]  = continuous_split_test
+        # everything else goes in train_mask
+        train_mask = 1 - test_mask
```
Comment (jtcramer) [inline sonyc_kalman/data.py]: An issue here (assuming I understand what's going on correctly) is that because we are creating the holes with respect to the valid indices, we risk splitting the holes with respect to the full set of indices. Consider this example with 10 samples
```
|0|1|2|3|4|5|6|7|8|9|
|-|-|-|-|X|-|-|-|-|-|
```
There are 9 valid indices. Say we sample a single hole from valid indices 2-4. The resulting train/valid split would be:

```
|0|1|2|3|4|5|6|7|8|9|
|T|T|V|V|T|V|T|T|T|T|
```

So we've split the validation hole into two, and we also have a training sequence that is a single invalid sample.

We can revisit doing this properly at some point later, but I think right now it's probably easier to just ignore the mask entirely when doing the sampling. Otherwise it becomes difficult trying to sample contiguous valid regions specifically for the validation set.

**Review by jtcramer — CHANGES_REQUESTED** [review]

File: sonyc_kalman/data.py (hunk 11 above)
Comment (tomxi) [reply sonyc_kalman/data.py]: Let me know your thoughts on 0bb2832

**Review by ChrisIck — APPROVED** [review]
LGTM

File: sonyc_kalman/data.py (hunk 11 above)
Comment (jtcramer) [reply sonyc_kalman/data.py]: LGTM

**Review by jtcramer — APPROVED** [review]
LGTM
